RFC-0001-torch-function-for-methods.md

```diff
@@ -0,0 +1,343 @@
+
+We propose to solve this problem with the following changes to PyTorch:
+
+1. Make methods and operators of `torch.Tensor` go through the
+   `__torch_function__` machinery.
+2. Add a `types` argument to `__torch_function__`, to make it match NumPy's
+   `__array_function__`.
+3. Make `torch.Tensor._make_subclass` public API by renaming it to `torch.Tensor.make_subclass`.
+4. Make `torch.Tensor` gain a generic implementation of `__torch_function__`.
+
+## Usage and Impact
+Once this proposal is merged, users of subclasses of `torch.Tensor` will have
+a much more streamlined experience. Namely, the following code example will
+work as-is, without the need for any further modification:
+
+```python
+class SubTensor(torch.Tensor):
+    a = 1
+
+t = SubTensor([1])
+s = t.sum()
+isinstance(s, SubTensor)  # True
+s.a  # 1
+i = t[0]
+isinstance(i, SubTensor)  # True
+i.a  # 1
+
+s2 = t + torch.Tensor(1)
+isinstance(s2, SubTensor)  # True
+s2.a  # 1
+
+s3 = torch.Tensor(1) + t
+isinstance(s3, SubTensor)  # True
+s3.a  # 1
+```
+
+Additionally, it will provide subclass authors the ability to also modify the
+results of methods and operators in `__torch_function__`, along with regular
+function calls, and to modify the result to their specific use-case, perform
+logging, or otherwise change the result or the action of the method. For
+example:
+
+```python
+import logging
+
+class LoggingTensor(torch.Tensor):
+    def __torch_function__(self, func, types, args, kwargs):
+        logging.info(f"func: {func.__name__}, args: {args!r}, kwargs: {kwargs!r}")
+        return super().__torch_function__(
+            func,
+            tuple(
+                t for t in types if not issubclass(t, LoggingTensor)
+            ),
+            args,
+            kwargs
+        )
+```
+
+Assuming minimum logging level is set to `logging.INFO`, the following
+indicates the code run, with the logging output in the comments.
+
+```python
+t = LoggingTensor([1])
+
+t.sum()  # sum, (LoggingTensor([1]),), {}
+t[0]  # __getitem__, (LoggingTensor([1]), 0,), {}
+
+# This is already possible
+torch.sum(t)  # sum, (LoggingTensor([1]),), {}
+```
+
+To make the protocol operate only on functions rather than methods, one can
+check for `func not in type(self).__dict__.values()`. To check for operators
+and/or indexing, one can check `func.__name__.endswith("__")`.
+
+### Performance
+There are a few requirements for the performance of this proposal, when
+implemented:
+
+1. No deterioration for function/method calls on `torch.Tensor` objects.
+2. No deterioration of current `__torch_function__` overhead
+3. Sub-µs impact on the performance of subclasses not implementing
+  `__torch_function__`.
+
+Requirement 1 seems unachievable due to the structure of the code at this
+point, as processing will have to be added for `self` in methods, as well as
+adding argument processing for methods where it doesn't already exist. We think
```

Comment:
Err, assuming that we have `__torch_function__` existing overhead, shouldn't there be no *further* overhead here? (Are you saying that the overhead from functions will then propagate to methods? We agreed that this overhead was fine for the `__torch_function__` patch.)

Reply:
Okay, there are two issues here:

1. The code that extracts out `Tensor`-likes for `__torch_function__` handling doesn't process `self`, and processing needs to be added for it. That will incur a minor cost.
2. For methods that only take `self` as an argument and no other `Tensor`-like, and are implemented in C++, those are skipped from the processing entirely, and processing needs to be added for them. That will incur a cost for these methods. These are methods like `tensor.sum()`.

Reply:
A third issue applies to methods defined in Python, to these the overhead will propagate.

Reply:
Added this to the proposal.